### PR TITLE
Pointer propagation fix for powerset-cherrypick

### LINKED ISSUE #2606: `CherryPick` rule for powersets incorrectly propagates pointers
See https://github.com/informalsystems/quint/issues/963

Any set picked from a powerset will currently hold the exact same pointers as the base set in the arena (though it _will_ be constrained in SMT). This leads to incorrect counterexample reconstruction (from a, presumably, correct model).

### PR DESCRIPTION
**Pull request by Kukovec**
- [ ] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

closes #2606

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala [hunk 1]
```diff
@@ -915,11 +915,24 @@ class CherryPick(rewriter: SymbStateRewriter) {
     val resultSet = nextState.arena.topCell
     val baseSet = nextState.arena.getDom(set)
     val elems = nextState.arena.getHasPtr(baseSet)
+
     // resultSet may contain all the elements from the baseSet of the powerset SUBSET(S)
-    nextState = nextState.updateArena(_.appendHas(resultSet, elems: _*))
+    // We have to restrict the pointers with base-set membership conditions
+    val (freshConsMap, restrictedElems) =
+      elems.foldLeft((Map.empty[ElemPtr, ArenaCell], Seq.empty[ElemPtr])) { case ((map, ptrs), elemPtr) =>
+        nextState = nextState.updateArena(_.appendCell(BoolT1))
```
Comment (konnov) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala:923]:
```suggestion
        // add a fresh Boolean constant
        nextState = nextState.updateArena(_.appendCell(BoolT1))
```

File: tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala [hunk 2]
```diff
@@ -915,11 +915,24 @@ class CherryPick(rewriter: SymbStateRewriter) {
...
+        val unconstrainedCell = nextState.arena.topCell
+        val ptr = elemPtr.restrict(unconstrainedCell.toBuilder)
+        val resPtrs = ptrs :+ ptr
```
Comment (konnov) [inline tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala:926]:
```suggestion
        // add the unconstrained edge to the list of pointers
        val resPtrs = ptrs :+ ptr
```

**Review by konnov — APPROVED** [review]
Looks like a right fix

**Comment by thpani** [comment]
Shouldn't this update the changelog, so we bump the version when cutting the release?

**Comment by Kukovec** [comment]
I didn't think this ticked "Entries added to `./unreleased/` _for any new functionality_`, but following the link, I guess bugfixes also qualify. We should update the PR template, "any new functionality" is imprecise.

**Pull request merged**